### PR TITLE
rangefeed: disable rangefeed memory budgets dynamically

### PR DESCRIPTION
**Pull request by aliher1911**
Previously memory budgets could only be disabled on new
rangefeeds where processor is not already running. Existing
rangefeeds will continue to limit used memory.
In case budgets needs to be urgently disabled on a big cluster
you'll have to restart all system rangefeeds which means you
likely end up restarting all nodes. This is disruptive and
should be avoided.
This PR adds an atomic flag that existing budgets could check
and stop allocation checks if it is reset to false.

Release note (ops change): Rangefeed memory budgets could be
disabled on the fly when cluster setting is changed without
the need to restart the feed.

### CODE REVIEW

File: pkg/kv/kvserver/rangefeed/budget.go [hunk 1]
```diff
@@ -118,16 +118,20 @@ type FeedBudget struct {
 	limit int64
 	// Channel to notify that memory was returned to the budget.
 	replenishC chan interface{}
-	// Budget cancellation request
+	// Budget cancellation request.
 	stopC chan interface{}
+	// Atomic value reflecting RangefeedBudgetsEnabled cluster setting. When
+	// budgets are disabled, all new messages should pass through, all previous
+	// allocations should still be returned as memory pool is shared with SQL.
+	budgetsEnabled *int32
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/rangefeed/budget.go:126]: I wonder though: is this any faster than simply checking the setting directly? That essentially does the same thing: checks an atomic int, with an additional map lookup.

https://github.com/cockroachdb/cockroach/blob/5b5fcee012ad2307e1b01748f44e865fd1a951f3/pkg/settings/values.go#L82

Comment (aliher1911) [reply pkg/kv/kvserver/rangefeed/budget.go:126]: Lookup in settings:
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkTryGet
BenchmarkTryGet-16      424617386            2.820 ns/op

BenchmarkTryGet
BenchmarkTryGet-16      441149536            2.649 ns/op

BenchmarkTryGet
BenchmarkTryGet-16      445762287            2.618 ns/op

Lookup via dedicated atomic:
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkTryGet
BenchmarkTryGet-16      440164717            2.642 ns/op

BenchmarkTryGet
BenchmarkTryGet-16      458073957            2.612 ns/op

BenchmarkTryGet
BenchmarkTryGet-16      455464820            2.598 ns/op

Looks like extra atomic is not worth the effort, we can just lookup directly and map lookup is fast.
Those are microbenchmarks, dunno if it'll have any effects on caches or anything, but we must be accessing settings pretty regularly so likely not too bad.

Comment (aliher1911) [reply pkg/kv/kvserver/rangefeed/budget.go]: Interestingly, adding a nil check for settings prior to reading them (to be able to not provide settings in tests) adds more to the bench time than switching from atomic to settings reads.

File: pkg/kv/kvserver/rangefeed/budget_test.go [hunk 2]
```diff
@@ -31,7 +31,10 @@ func TestFeedBudget(t *testing.T) {
 		m.Start(context.Background(), nil, mon.MakeStandaloneBudget(poolSize))
 		b := m.MakeBoundAccount()
 
-		f := NewFeedBudget(&b, budgetSize)
+		//enabled := int32(1)
+		//f := NewFeedBudget(&b, budgetSize, &enabled)
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/rangefeed/budget_test.go:35]: nit: stray comments?

File: pkg/kv/kvserver/rangefeed/processor_test.go [hunk 3]
```diff
@@ -1086,10 +1081,12 @@ func TestBudgetReleaseOnProcessorStop(t *testing.T) {
 	// as sync events used to flush queues.
 	const channelCapacity = totalEvents/2 + 10
 
+	s := cluster.MakeTestingClusterSettings()
 	m := mon.NewMonitor("rangefeed", mon.MemoryResource, nil, nil, 1, math.MaxInt64, nil)
 	m.Start(context.Background(), nil, mon.MakeStandaloneBudget(math.MaxInt64))
+	//budgetEnabled := int32(1)
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/rangefeed/processor_test.go:1087]: nit: stray comment.

**Review by erikgrinaker — APPROVED** [review]

File: pkg/kv/kvserver/rangefeed/budget.go (hunk 1 above)
Comment (erikgrinaker) [reply pkg/kv/kvserver/rangefeed/budget.go]: Nice, that's what I thought -- direct settings lookup is fine here.

I've also been surprised by how expensive those `nil` checks can be, I've found an extra `bool` to signal presence to be much more performant. Anyway, requiring settings in tests seems fine.

File: pkg/kv/kvserver/rangefeed/budget_test.go (hunk 2 above)
Comment (aliher1911) [reply pkg/kv/kvserver/rangefeed/budget_test.go:35]: Ah, leftovers of the benchmarking!

**Comment by aliher1911** [comment]
bors r=erikgrinaker

**Pull request merged**